src/main.cpp

```diff
@@ -1,6 +1,227 @@
+#include <string_view>
+#include <unordered_map>
+
+#define DEBUG_MODE /* empty */
+
+using tl_Number = double;
+
+enum class TokenType {
+	plus,	// +
+	minus,	// -
+	mult,	// *
+	div,	// /
+	mod,	// %
+	num,	// [0-9]+(.[0-9]+)?
+	string, // '[^']' | "[^"]"
+	true_,	// true
+	false_, // false
+	let,	// let
+	nil,	// nil
+	eof_,	// '/0'
+	error,	// unknown
+	eCOUNT_
+};
+
+using TT = TokenType;
+
+// clang-format off
+static constexpr std::array<std::string_view, (std::size_t)TT::eCOUNT_> tokenNames = {
+	"PLUS",	
+	"MINUS",	
+	"MULT",	
+	"DIV",	
+	"MOD",	
+	"NUM",	
+	"STRING", 
+	"TRUE",	
+	"FALSE",	
+	"LET",	
+	"NIL",
+	"END_OF_FILE",	
+	"<ERROR>"
+};
+// clang-format on
+
+const static std::unordered_map<std::string_view, TokenType> keywords = {
+	// clang-format off
+	{"nil"  , TT::nil   },
+	{"let"  , TT::let   },
+	{"true" , TT::true_ },
+	{"false", TT::false_},
+	// clang-format on
+};
+
+struct TokenLocation {
+	std::size_t start = 0;
+	std::size_t length = 0;
+	std::size_t line = 0;
+	std::uint32_t column = 0;
+
+	std::string_view view(const std::string& source) const {
+		return std::string_view(source.c_str() + start, length);
+	}
+};
+
+struct Token {
+	TT type = TT::error;
+
+	TokenLocation location = {};
+
+	std::string_view raw(const std::string& source) const {
+		return location.view(source);
+	}
+};
+
+std::ostream& operator<<(std::ostream& os, const Token& token) {
+	return os << tokenNames[(std::size_t)(token.type)];
+}
+
+class Lexer {
+  private:
+	struct SourcePosition {
+		std::size_t current = 0;
+		std::size_t start = 0;
+		std::size_t line = 1;
+		std::uint32_t column = 0;
+	} pos;
+
+	std::string_view fileName;
+	std::string source;
+
+  public:
+	Lexer(std::string_view& fileName, std::string& src)
+		: fileName(fileName), source(src) {
+	}
+
+	Token nextToken() {
+		skipWhiteSpace();
+		pos.start = pos.current;
+
+		if (eof()) {
+			return makeToken(TT::eof_);
+		}
+		
+		char c = advance();
+
+		// clang-format off
+		switch (c) {
+			case '+': return makeToken(TT::plus );
+			case '-': return makeToken(TT::minus);
+			case '*': return makeToken(TT::mult );
+			case '/': return makeToken(TT::div  );
+			case '%': return makeToken(TT::mod  );
+			default:  
+			if (isdigit(c)) return number();
+		}
+		// clang-format on
+		return makeToken(TT::error);
+	}
+
+  private:
+	Token number() {
+		while (isdigit(peek())) {
+			advance();
+		}
+
+		if (peek() == '.' && isdigit(peekNext())) {
+			advance(); // consume the '.'
+			while (isdigit(peek())) {
+				advance();
+			}
+		}
+
+		return makeToken(TT::num);
+	}
+
+	// helper functions:
+
+	// somewhat expensive, optimize later.
+	bool isKeyword(Token& token) const {
+		return (keywords.find(token.raw(source)) != keywords.end());
+	}
+
+	char advance() {
+		if (pos.current >= source.length())
+			return '\0';
+		return source[pos.current++];
+	}
+
+	char peek() const {
+		if (eof())
+			return '\0';
+		return source[pos.current];
+	}
+
+	char peekNext() const {
```

Comment:
This could instead be `peekAhead(std::size_t dx = 1)`

Reply:
regular grammars concerning regexes only require a single character lookAhead, should I still implement the `dx` argument ?